In [1]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

from warnings import filterwarnings

from IPython.display import display, Markdown, Latex

filterwarnings('ignore')

In [2]:
dftrain = pd.read_csv('train.csv',)
dftest = pd.read_csv('test_X.csv', header=0, sep=';')

### Primera vista de los datos

In [3]:
dftrain.head()

order_id local_time country_code  store_address payment_status  \
0  33446280   14:11:09           AR          55379           PAID   
1  33107339   11:47:41           GT          23487           PAID   
2  32960645   11:53:53           CR          62229           PAID   
3  32089564   20:15:21           ES          29446           PAID   
4  32157739   21:32:16           AR          13917           PAID   

   n_of_products  products_total     final_status  
0              2           11.88  DeliveredStatus  
1              2            5.20  DeliveredStatus  
2              1            6.03  DeliveredStatus  
3              6            6.37  DeliveredStatus  
4              1            5.36   CanceledStatus

In [4]:
dftest

order_id local_time country_code  store_address payment_status  \
0   32233784   17:50:09           MA          68169           PAID   
1   32240990   18:38:08           ES           8220           PAID   
2   33331821   22:11:59           IT          11169           PAID   
3   33200505   22:13:55           AR          33371           PAID   
4   32527480   12:01:04           TR          33958           PAID   
5   32054504   11:15:39           EC          51386           PAID   
6   33447533   18:18:18           ES          16351           PAID   
7   32238371   20:21:09           TR          49659           PAID   
8   32960491   18:56:17           MA          32273       NOT_PAID   
9   32430153   13:43:26           PA          17782           PAID   
10  32777553   09:03:00           TR          28205           PAID   
11  32324328   21:15:20           AR          64309           PAID   
12  33047576   11:42:08           UA          61562           PAID   
13  33293677   20:03:33           ES          22022           PAID   
14  32073867   14:12:57           DO          48440           PAID   
15  32732256   21:31:36           ES          18667           PAID   
16  32152540   18:58:35           EC          41406           PAID   
17  32628045   20:35:24           AR          70877           PAID   
18  33226955   13:00:39           EG          50260           PAID   
19  33559535   22:53:51           EC          70373           PAID   
20  32400294   16:58:39           ES          21625           PAID   
21  33317211   15:21:15           PE          51279           PAID   
22  33475314   14:10:55           EC          54679           PAID   
23  33043612   10:42:03           TR          50632           PAID   
24  32968657   20:37:52           UA          64308           PAID   
25  32160619   21:51:05           AR          21110           PAID   
26  33435482   16:55:31           ES          47795           PAID   
27  33336368   22:32:02           FR           3705           PAID   
28  33180028   17:44:53           PA          18746           PAID   
29  32849661   20:33:41           IT          10369           PAID   

    n_of_products  products_total  
0               1           61.63  
1              11           15.99  
2               4            5.89  
3               3            7.85  
4               2            4.75  
5               2           14.28  
6               5           14.35  
7               2            4.42  
8               2            1.79  
9               5           29.89  
10              1            1.47  
11              1            4.20  
12              6           18.73  
13              1            3.80  
14              1            6.25  
15              2           14.50  
16              1           21.10  
17              2            9.85  
18              1            0.77  
19              1            2.89  
20              1            4.30  
21              4           26.22  
22              1            4.44  
23              2            3.35  
24              2            7.57  
25              1            0.89  
26             10           20.40  
27              2            4.95  
28              1            7.58  
29              6           53.00

In [5]:
dftrain.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54330 entries, 0 to 54329
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   order_id        54330 non-null  int64  
 1   local_time      54330 non-null  object 
 2   country_code    54330 non-null  object 
 3   store_address   54330 non-null  int64  
 4   payment_status  54330 non-null  object 
 5   n_of_products   54330 non-null  int64  
 6   products_total  54330 non-null  float64
 7   final_status    54330 non-null  object 
dtypes: float64(1), int64(3), object(4)
memory usage: 3.3+ MB


In [6]:
dftest.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30 entries, 0 to 29
Data columns (total 7 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   order_id        30 non-null     int64  
 1   local_time      30 non-null     object 
 2   country_code    30 non-null     object 
 3   store_address   30 non-null     int64  
 4   payment_status  30 non-null     object 
 5   n_of_products   30 non-null     int64  
 6   products_total  30 non-null     float64
dtypes: float64(1), int64(3), object(3)
memory usage: 1.8+ KB


#### Variables

- Numéricas: 
    - n_of_products
    - products_total 

- Categóricas: 
    - country_code
    - store_address
    - payment_status  

- Fecha:
    - local_time

- Variable dependiente:
    - final_status

In [7]:
dftrain['local_time'] = pd.to_datetime(dftrain.local_time)
dftest['local_time'] = pd.to_datetime(dftest.local_time)

### 1. ¿Cuáles son los 3 paises en los que más pedidos se realizan?



In [8]:
dftrain.groupby(by='country_code').store_address.count().sort_values(ascending=False).head(3)

country_code
AR    11854
ES    11554
TR     5696
Name: store_address, dtype: int64

In [9]:
NUM_PAISES = 3
data = dftrain.groupby(by='country_code').store_address.count().sort_values(ascending=False).reset_index().head(NUM_PAISES)
md = "|Pais|Pedidos|\n|:-|-:|\n"
for i in range(3):
    md = md + f"|{data.loc[i,'country_code']}|{data.loc[i,'store_address']}|\n"
Markdown(md)

|Pais|Pedidos|
|:-|-:|
|AR|11854|
|ES|11554|
|TR|5696|


### 2. ¿Cuáles son las horas en las que se realizan más pedidos en España?


In [10]:
mask = dftrain.country_code == 'ES'

data = dftrain[mask].groupby(pd.Grouper(key='local_time',freq='h'))\
                .store_address.count()\
                .sort_values(ascending=False)\
                .reset_index()\

data['local_time'] = data.local_time.dt.strftime("%H:%M")

data.head()

local_time  store_address
0      20:00           1716
1      21:00           1155
2      19:00           1128
3      13:00           1047
4      14:00            956

### 3. ¿Cuál es el precio medio por pedido en la tienda con ID 12513? 


In [11]:
mask = dftrain.store_address == 12513
val = dftrain[mask].products_total.mean()

Markdown(f"El precio medio por pedido en la tienda es **{val.round(4)}**")


El precio medio por pedido en la tienda es **17.3922**

### 4. Teniendo en cuenta los picos de demanda en España, si los repartidores trabajan en turnos de 8horas.  
Turno 1 (00:00-08:00)  
Turno 2 (08:00-16:00)  
Turno 3 (16:00-00:00)  
Qué porcentaje de repartidores pondrías por cada turno para que sean capaces de hacer frente a los picos de demanda. (ej: Turno 1 el 30%, Turno 2 el 10% y Turno 3 el 60%).

In [12]:
PAIS = 'ES'
mask = dftrain.country_code == PAIS
data = dftrain[mask].groupby(pd.Grouper(key='local_time',freq='8h')).store_address.count() / dftrain[mask].store_address.count()
data = data.reset_index()
data['local_time'] = data.local_time.dt.strftime("%H:%M")

data

local_time  store_address
0      00:00       0.001212
1      08:00       0.388350
2      16:00       0.610438

In [13]:
md = "##### Porcentaje de repartidores por cada turno  \n"
md = md + "|Turno|% repartidores|\n|:-|-:|\n"
for i in range(3):
    md = md + f"|{data.loc[i,'local_time']}|{(data.loc[i,'store_address']*100).round(3)}%|\n"
Markdown(md)

##### Porcentaje de repartidores por cada turno  
|Turno|% repartidores|
|:-|-:|
|00:00|0.121%|
|08:00|38.835%|
|16:00|61.044%|


Considerar que hablamos de todo el territorio, con lo que, teniendo en cuenta los turnos, el número de tiendas y que desconocemos cuantos repartidores hay, estos valores son aproximados (y hay que considerar un mínimo de repartidores en las tiendas que tienen actividad en el primer turno)

In [14]:
dftrain[mask].groupby(pd.Grouper(key='local_time',freq='8h')).store_address.nunique() 

local_time
2021-11-28 00:00:00      4
2021-11-28 08:00:00    679
2021-11-28 16:00:00    732
Freq: 8H, Name: store_address, dtype: int64

In [15]:
Markdown(f"Número de tiendas en {PAIS}: **{dftrain[mask].store_address.nunique()}**")

Número de tiendas en ES: **822**

Hay 4 tiendas que han repartido en el Turno 1, del total de 822, un 0.4% del total, y han servido el 0.12% de pedidos del dia

### 5. Realiza un modelo predictivo de machine learning
 ... a partir del dataset 'train.csv' en el cual a partir de las variables predictoras que se entregan en el dataset 'test_X' se pueda predecir si el pedido se cancelará o no (columna 'final_status'). Siendo:
Para simplificar, podeis asignar los valores 'CanceledStatus' a 0 y los valores 'DeliveredStatus' a 1.

> 0 = CanceledStatus  
> 1 = DeliveredStatus

In [16]:
from sklearn.model_selection import StratifiedKFold, train_test_split
from sklearn.preprocessing import MinMaxScaler, PowerTransformer, RobustScaler
from sklearn.metrics import f1_score, confusion_matrix
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import OneHotEncoder, LabelEncoder
from sklearn.compose import ColumnTransformer

from imblearn.over_sampling import KMeansSMOTE, SMOTE
from category_encoders import TargetEncoder, OneHotEncoder as OHE

from xgboost import XGBClassifier
from catboost import CatBoostClassifier
from lightgbm import LGBMClassifier
from sklearn.svm import SVC
from sklearn.linear_model import LogisticRegression


from tqdm.notebook import tqdm


import time




In [17]:
SEED = 42
TEST_SIZE = 0.2
TESTING = True

In [18]:
def make_pipeline(data, columns, model):
    cols_num = data[columns].select_dtypes(include=['int','float']).columns.to_list()
    cols_cat = data[columns].select_dtypes(include=['object']).columns.to_list()
    cols_date = data[columns].select_dtypes(include=['datetime']).columns.to_list()
    
    ct = ColumnTransformer(
             transformers=[('num', MinMaxScaler(), cols_num),
                          ('cat',OneHotEncoder(handle_unknown='ignore'), cols_cat)])
    
    pipeline = Pipeline( steps=[('transformer', ct),
                               ('clasif', model)])
    
    return pipeline


def run_experiment(train, test=None, columns=[], model=LogisticRegression()):
    if test is None:
        X_train, X_test, y_train, y_test = train_test_split(*train,  
                                                        test_size=TEST_SIZE, random_state=SEED)
    else:
        X_train, y_train = train
        X_test, y_test = test
        
#     pipe = make_pipeline(X_train, columns, model)
    
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    
    return f1_score(y_test, np.round(y_pred), average='macro'), model
  

### Feature Engineering and Preprocessing

In [19]:
# Mapeamos 

# target_map = {k:v for v,k in enumerate(dftrain.final_status.unique())}

target_map = {'CanceledStatus': 0, 'DeliveredStatus': 1}


dftrain['status'] = dftrain.final_status.map(target_map)

In [20]:
dftrain['hour'] = dftrain.local_time.dt.hour.astype(int)
dftest['hour'] = dftest.local_time.dt.hour.astype(int)


In [21]:
daily_del_list = dftrain.groupby('store_address').order_id.count().to_dict()
dftrain['daily_del'] = dftrain.store_address.map(daily_del_list)
dftest['daily_del'] = dftrain.store_address.map(daily_del_list)

In [22]:
from category_encoders import TargetEncoder

te = TargetEncoder()
dftrain['store_enc'] = te.fit_transform(dftrain['store_address'],dftrain.status)
dftest['store_enc'] = te.transform(dftest['store_address'])


dftrain['time_shift'] = dftrain.local_time.dt.floor('8h')
dftest['time_shift'] = dftest.local_time.dt.floor('8h')

dftrain['shift_del'] = dftrain.groupby(by=['store_address','time_shift']).order_id.transform('count')

In [23]:
dftest = dftest.merge(dftrain[['store_address','time_shift','shift_del']].drop_duplicates(), \
             on=['store_address','time_shift'], how='left')
dftest.shift_del.fillna(0.0, inplace=True)

In [24]:
dftest

order_id          local_time country_code  store_address payment_status  \
0   32233784 2021-11-28 17:50:09           MA          68169           PAID   
1   32240990 2021-11-28 18:38:08           ES           8220           PAID   
2   33331821 2021-11-28 22:11:59           IT          11169           PAID   
3   33200505 2021-11-28 22:13:55           AR          33371           PAID   
4   32527480 2021-11-28 12:01:04           TR          33958           PAID   
5   32054504 2021-11-28 11:15:39           EC          51386           PAID   
6   33447533 2021-11-28 18:18:18           ES          16351           PAID   
7   32238371 2021-11-28 20:21:09           TR          49659           PAID   
8   32960491 2021-11-28 18:56:17           MA          32273       NOT_PAID   
9   32430153 2021-11-28 13:43:26           PA          17782           PAID   
10  32777553 2021-11-28 09:03:00           TR          28205           PAID   
11  32324328 2021-11-28 21:15:20           AR          64309           PAID   
12  33047576 2021-11-28 11:42:08           UA          61562           PAID   
13  33293677 2021-11-28 20:03:33           ES          22022           PAID   
14  32073867 2021-11-28 14:12:57           DO          48440           PAID   
15  32732256 2021-11-28 21:31:36           ES          18667           PAID   
16  32152540 2021-11-28 18:58:35           EC          41406           PAID   
17  32628045 2021-11-28 20:35:24           AR          70877           PAID   
18  33226955 2021-11-28 13:00:39           EG          50260           PAID   
19  33559535 2021-11-28 22:53:51           EC          70373           PAID   
20  32400294 2021-11-28 16:58:39           ES          21625           PAID   
21  33317211 2021-11-28 15:21:15           PE          51279           PAID   
22  33475314 2021-11-28 14:10:55           EC          54679           PAID   
23  33043612 2021-11-28 10:42:03           TR          50632           PAID   
24  32968657 2021-11-28 20:37:52           UA          64308           PAID   
25  32160619 2021-11-28 21:51:05           AR          21110           PAID   
26  33435482 2021-11-28 16:55:31           ES          47795           PAID   
27  33336368 2021-11-28 22:32:02           FR           3705           PAID   
28  33180028 2021-11-28 17:44:53           PA          18746           PAID   
29  32849661 2021-11-28 20:33:41           IT          10369           PAID   

    n_of_products  products_total  hour  daily_del  store_enc  \
0               1           61.63    17          4      68169   
1              11           15.99    18         21       8220   
2               4            5.89    22         10      11169   
3               3            7.85    22         39      33371   
4               2            4.75    12         51      33958   
5               2           14.28    11        145      51386   
6               5           14.35    18          4      16351   
7               2            4.42    20         20      49659   
8               2            1.79    18         66      32273   
9               5           29.89    13         13      17782   
10              1            1.47     9         57      28205   
11              1            4.20    21         37      64309   
12              6           18.73    11          5      61562   
13              1            3.80    20         26      22022   
14              1            6.25    14         70      48440   
15              2           14.50    21         29      18667   
16              1           21.10    18          1      41406   
17              2            9.85    20        111      70877   
18              1            0.77    13         40      50260   
19              1            2.89    22         25      70373   
20              1            4.30    16         43      21625   
21              4           26.22    15         10      51279   
22              1            4.44    14         15      546

In [25]:
cat_cols = ['country_code','payment_status']
num_cols = ['n_of_products', 'products_total', 'store_enc','hour','daily_del', 'shift_del']
date_cols = ['local_time']

train_cols = cat_cols + num_cols
target = ['status']

In [26]:

FOLDS = 7
dftrain['fold'] = -1

skf = StratifiedKFold(n_splits=FOLDS, random_state=42, shuffle=True)

for fold, (train_idx, val_idx) in enumerate(skf.split(dftrain[train_cols], dftrain[target])):
    print(fold, train_idx.shape, val_idx.shape)
    dftrain.loc[val_idx, 'fold'] = fold

0 (46568,) (7762,)
1 (46568,) (7762,)
2 (46568,) (7762,)
3 (46569,) (7761,)
4 (46569,) (7761,)
5 (46569,) (7761,)
6 (46569,) (7761,)


In [27]:
preds_oof = []
final_val_preds = {}

for fold in range(FOLDS):
    xtrain = dftrain[dftrain.fold != fold]
    xval = dftrain[dftrain.fold == fold]
    xtest= dftest
    
    valid_ids = xval.index.to_list()
    
    ytrain = xtrain[target]
    yval = xval[target]
    
    params = {'learning_rate': 0.04113078629577159,
             'reg_lambda': 2.2529987562292753e-05,
             'reg_alpha': 2.1553668505558285e-06,
             'subsample': 0.5079109680078884,
             'colsample_bytree': 0.8912508896212556,
             'max_depth': 7,
#              'objective': 'binary:logistic',
             'metrics': 'f1'}
    
    xtrain = xtrain[train_cols]
    xval = xval[train_cols]
    xtest = xtest[train_cols]
    
    model = LGBMClassifier(is_unbalance=True, **params)
#     model = CatBoostClassifier(class_weights=[0.1, 0.9])
        
    ct = ColumnTransformer(
             transformers=[('num', RobustScaler(), num_cols),
                          ('cat',OneHotEncoder(handle_unknown='ignore'), cat_cols)
                          ])
    
    xtrain_tr = ct.fit_transform(xtrain)
    xval_tr = ct.transform(xval)
    xtest_tr = ct.transform(xtest)
    
    model.fit(xtrain_tr, ytrain, eval_set=[(xtrain_tr,ytrain),(xval_tr,yval)], verbose=0 )
    
    preds_valid = model.predict_proba(xval_tr)
    preds_oof.append(model.predict_proba(xtest_tr)[:,1])
    final_val_preds.update(dict(zip(valid_ids, preds_valid)))
    f1 = f1_score(yval, model.predict(xval_tr), average='macro')
    f1tr = f1_score(ytrain, model.predict(xtrain_tr), average='macro')
    print(f'Fold {fold}\t val f1: {f1:.5f}, train f1: {f1tr:.5f} ')
    

Fold 0	 val f1: 0.55061, train f1: 0.57581 
Fold 1	 val f1: 0.55295, train f1: 0.57312 
Fold 2	 val f1: 0.55107, train f1: 0.57384 
Fold 3	 val f1: 0.55579, train f1: 0.57434 
Fold 4	 val f1: 0.55790, train f1: 0.57334 
Fold 5	 val f1: 0.55191, train f1: 0.57321 
Fold 6	 val f1: 0.55865, train f1: 0.57684 


In [28]:
np.unique(np.array(preds_oof).mean(axis=0).round(), return_counts=True)

(array([0., 1.]), array([ 7, 23]))

In [29]:
dfpred = pd.DataFrame(np.array(preds_oof).mean(axis=0).round(), columns=['final_status'])
dfpred.to_csv('predicciones.csv')